File: exam_LRX/cpp/work2.cpp
//
// Created by liryi on 24-11-7.
//2.提取图像中的叶片病害的图斑数目、估算病害图斑占叶片的总面比例。

#include <opencv2/opencv.hpp>
#include <iostream>

int main() {
    // 读取图像
    cv::Mat image = cv::imread("E:\\DL_class\\exam_LRX\\cpp\\images\\image2.png");
    if (image.empty()) {
        std::cout << "Could not open or find the image!" << std::endl;
        return -1;
    }

    // 将图像转换为 HSV 颜色空间
    cv::Mat image_hsv;
    cv::cvtColor(image, image_hsv, cv::COLOR_BGR2HSV);

    // 定义黄色区域的 HSV 范围
    cv::Scalar lower_yellow(15, 80, 80);  // 放宽范围以捕捉更多黄色区域
    cv::Scalar upper_yellow(35, 255, 255);
    cv::Mat mask_yellow;
    cv::inRange(image_hsv, lower_yellow, upper_yellow, mask_yellow);

    // 对掩膜进行形态学操作（使用较小的核）
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(3, 3));
    cv::morphologyEx(mask_yellow, mask_yellow, cv::MORPH_CLOSE, kernel);

    // 计算叶片总面积和病变面积
    cv::Mat gray;
    cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);
    int leaf_area = cv::countNonZero(gray);
    int disease_area = cv::countNonZero(mask_yellow);
    double disease_ratio = (static_cast<double>(disease_area) / leaf_area) * 100.0;

    // 输出结果
    std::cout << "Disease Area Ratio: " << disease_ratio << "%" << std::endl;

    cv::Mat mask_yellow_color;
    cv::cvtColor(mask_yellow, mask_yellow_color, cv::COLOR_GRAY2BGR);

    // 拼接原图和掩膜图像
    cv::Mat combined;
    cv::hconcat(image, mask_yellow_color, combined);

    // 显示拼接后的图像
    cv::imshow("Original and Disease Mask", combined);
    cv::waitKey(0);

    return 0;

}

File: exam_latest/exam_CMM/work2.py
# 2.提取图像中的叶片病害的图斑数目、估算病害图斑占叶片的总面比例。
# -*- coding: utf-8 -*-
import cv2
import numpy as np
import matplotlib.pyplot as plt

# 读取图像
image = cv2.imread('image/image2.png')
image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

# 将图像从RGB转换为HSV
image_hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)

# 设置黄色病变区域的HSV颜色范围
lower_yellow = np.array([20, 100, 100])
upper_yellow = np.array([30, 255, 255])
mask_yellow = cv2.inRange(image_hsv, lower_yellow, upper_yellow)

# 对掩膜应用形态学操作以去除噪声
kernel = np.ones((3, 3), np.uint8)
mask_yellow = cv2.morphologyEx(mask_yellow, cv2.MORPH_CLOSE, kernel)

# 计算病变面积占总面积的比例
leaf_area = np.count_nonzero(cv2.cvtColor(image, cv2.COLOR_BGR2GRAY) > 0)
disease_area = np.count_nonzero(mask_yellow)
disease_ratio = (disease_area / leaf_area) * 100

# 显示结果
plt.subplot(1, 2, 1)
plt.title("Original Image")
plt.imshow(image_rgb)

plt.subplot(1, 2, 2)
plt.title(f"Disease Area (Ratio: {disease_ratio:.2f}%)")
plt.imshow(mask_yellow, cmap="gray")

plt.show()

# 代码逻辑分析
# 图像读取与预处理：读取图像并将其转换为 HSV 色彩空间，用于分割特定颜色范围的病变区域。
# 病害区域检测：定义了黄色病变区域的 HSV 范围（lower_yellow 和 upper_yellow），通过掩膜提取病害区域。
# 形态学操作：使用闭操作清理噪声并填补病害区域的间隙，确保病害区域连贯。
# 面积计算：总叶片面积：通过灰度转换提取叶片有效区域的像素数。
# 病害面积：计算掩膜中非零像素数。
# 病害比例：病害面积占总叶片面积的百分比。
# 结果显示：使用 matplotlib 将原始图像和病害掩膜直观对比，并标注病害比例